### DISCUSSION TITLE
Error after Upgrade "<class 'django.db.utils.ProgrammingError'>"

### DISCUSSION THREAD

**Discussion by KleepX** [netbox-community/netbox#15800] [Help Wanted!]
Hello,

I am currently using Netbox version 3.2.1 and wanted to upgrade to the latest version 3.7.5 today. Unfortunately, after the upgrade I keep getting the error "<class 'django.db.utils.ProgrammingError'>" 

My Python library is on 3.8.10 and my Postgres database is on version 12, so there should be no problem.

The upgrade goes through without any problems, but after that I can no longer access the interface. Am I missing something?

![image](https://github.com/netbox-community/netbox/assets/61327547/96aa9156-d77a-4f65-a7d1-87a644d18b02)

The only thing the log tells me is:
```
Apr 22 13:26:39 netbox systemd[2139614]: netbox.service: Failed to determine user credentials: No such process
Apr 22 13:26:39 netbox systemd[2139614]: netbox.service: Failed at step USER spawning /opt/netbox/venv/bin/gunicorn: No such process
Apr 22 13:26:39 netbox systemd[1]: netbox.service: Main process exited, code=exited, status=217/USER
Apr 22 13:26:39 netbox systemd[1]: netbox.service: Failed with result 'exit-code'.
Apr 22 13:27:09 netbox systemd[1]: netbox.service: Scheduled restart job, restart counter is at 2.
Apr 22 13:27:09 netbox systemd[1]: Stopped NetBox WSGI Service.
Apr 22 13:27:09 netbox systemd[1]: Started NetBox WSGI Service.
Apr 22 13:27:09 netbox systemd[2139638]: netbox.service: Failed to determine user credentials: No such process
Apr 22 13:27:09 netbox systemd[2139638]: netbox.service: Failed at step USER spawning /opt/netbox/venv/bin/gunicorn: No such process
Apr 22 13:27:09 netbox systemd[1]: netbox.service: Main process exited, code=exited, status=217/USER
Apr 22 13:27:09 netbox systemd[1]: netbox.service: Failed with result 'exit-code'.
Apr 22 13:27:40 netbox systemd[1]: netbox.service: Scheduled restart job, restart counter is at 3.
Apr 22 13:27:40 netbox systemd[1]: Stopped NetBox WSGI Service.
Apr 22 13:27:40 netbox systemd[1]: Started NetBox WSGI Service.
Apr 22 13:27:40 netbox systemd[2139657]: netbox.service: Failed to determine user credentials: No such process
Apr 22 13:27:40 netbox systemd[2139657]: netbox.service: Failed at step USER spawning /opt/netbox/venv/bin/gunicorn: No such process
Apr 22 13:27:40 netbox systemd[1]: netbox.service: Main process exited, code=exited, status=217/USER
Apr 22 13:27:40 netbox systemd[1]: netbox.service: Failed with result 'exit-code'.
Apr 22 13:28:10 netbox systemd[1]: netbox.service: Scheduled restart job, restart counter is at 4.
Apr 22 13:28:10 netbox systemd[1]: Stopped NetBox WSGI Service.
Apr 22 13:28:10 netbox systemd[1]: Started NetBox WSGI Service.
Apr 22 13:28:10 netbox systemd[2139676]: netbox.service: Failed to determine user credentials: No such process
Apr 22 13:28:10 netbox systemd[2139676]: netbox.service: Failed at step USER spawning /opt/netbox/venv/bin/gunicorn: No such process
Apr 22 13:28:10 netbox systemd[1]: netbox.service: Main process exited, code=exited, status=217/USER
Apr 22 13:28:10 netbox systemd[1]: netbox.service: Failed with result 'exit-code'.
``` 

Edit:
I already tried to manually execute the "manage.py" script and got the error:

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py", line 395, in execute
    django.setup()
  File "/usr/local/lib/python3.8/dist-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.8/dist-packages/django/apps/registry.py", line 91, in populate
    app_config = AppConfig.create(entry)
  File "/usr/local/lib/python3.8/dist-packages/django/apps/config.py", line 224, in create
    import_module(entry)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'drf_spectacular'
```

**Comment by candlerb** [comment]
> I already tried to manually execute the "manage.py" script and got the error:

If migrations failed, then you would be trying to run Netbox 3.7.x code with a Netbox 3.2.x database, and that definitely won't work.

However, I also note from your screenshot (which I can't copy-paste) that it says NetBox version: 3.2.1.  This suggests that no upgrade took place at all.

There are two different upgrade procedures [documented](https://docs.netbox.dev/en/stable/installation/upgrading/) - one for when Netbox was originally installed from git, and one for when it was originally installed from tarball. It's important that you use the correct one.

If you're not sure which, use `ls -l /opt/netbox` and `ls -l /opt/netbox/.git` to check.

>  No module named 'drf_spectacular'

You're running manage.py wrongly, without activating the virtualenv first.

Run it like this instead and it should be fine:

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py ...
```

However: when you ran `/opt/netbox/upgrade.sh` it would have taken care of this for you. Presumably, then, there was some error when you ran this - can you show what it is? You can re-run it safely (it will rebuild your virtualenv first)

> /usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py

You should *not* be using any globally-installed packages for Django or any other python libraries. The packages will all be installed in Netbox's virtualenv instead.

**Reply by KleepX** [reply]
Hey Candlerb,

thank you for your Response. Soo i took a look. The installation seems to have been over git so i tried the Git way again like in the Documentation. 
```
You are installing (or upgrading to) NetBox version 3.7.6
Using Python 3.8.10
Removing old virtual environment...
Creating a new virtual environment at /opt/netbox/venv...
Updating pip (pip install --upgrade pip)...
Collecting pip
  Using cached pip-24.0-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.0.2
    Uninstalling pip-20.0.2:
      Successfully uninstalled pip-20.0.2
Successfully installed pip-24.0
pip 24.0 from /opt/netbox/venv/lib/python3.8/site-packages/pip (python 3.8)
Installing Python system packages (pip install wheel)...
Collecting wheel
  Using cached wheel-0.43.0-py3-none-any.whl.metadata (2.2 kB)
Using cached wheel-0.43.0-py3-none-any.whl (65 kB)
Installing collected packages: wheel
[...]
``` 

The upgrade Script ran without any Problems and Errors. So there shouldnt be any issues.

But is it a problem if i update and start the service as the Root user? Because in the Systemd logs there are errors that the USER cant be spawned.

```
netbox.service: Failed at step USER spawning /opt/netbox/venv/bin/gunicorn: No such process
```

> Run it like this instead and it should be fine:

> /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py ...

I tried it after the upgrade this is the Result:

```
root@netbox:/opt/netbox/contrib# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py migrate
Operations to perform:
  Apply all migrations: account, admin, auth, circuits, contenttypes, core, dcim, django_rq, extras, ipam, sessions, social_django, taggit, tenancy, users, virtualization, vpn, wireless
Running migrations:
  No migrations to apply.
```

**Reply by candlerb** [reply]
> But is it a problem if i update and start the service as the Root user? Because in the Systemd logs there are errors that the USER cant be spawned.

Does `/opt/netbox/venv/bin/gunicorn` exist? If not, then likely you're not building under the default prefix `/opt/netbox`. Show the exact command you typed, and the working directory you were in at the time, as well as its output. e.g.

```
pwd
./upgrade.sh
```
